src/GovUk.Education.ExploreEducationStatistics.Admin/Services/UserPublicationRoleAndInviteManager.cs

```
@@ -0,0 +1,132 @@
+    public async Task RemoveManyRolesAndInvites(
+        IReadOnlyList<UserPublicationRole> userPublicationRoles,
+        CancellationToken cancellationToken = default)
+    {
+        var inviteKeys = userPublicationRoles
+            .Select(upr => new
+            {
+                upr.PublicationId,
+                upr.Role,
+                Email = upr.User.Email.ToLower()
+            })
+            .ToList();
+
+        var invites = await ContentDbContext.UserPublicationInvites
+            .Where(upi => inviteKeys
+                .Contains(new
+                {
+                    upi.PublicationId,
+                    upi.Role,
+                    Email = upi.Email.ToLower()
+                }))
+            .ToListAsync(cancellationToken);
+
+        await ContentDbContext.RequireTransaction(async () =>
+        {
+            // Remove all of the associated invites if there are any
+            await userPublicationInviteRepository.RemoveMany(invites, cancellationToken);
+
+            await RemoveMany(userPublicationRoles, cancellationToken);
+        });
+    }
+
+    public async Task RemoveAllRolesAndInvitesForUser(
+        Guid userId,
+        CancellationToken cancellationToken = default)
+    {
+        var userEmail = await GetUserEmail(userId, cancellationToken);
+
+        var userPublicationRoles = await ContentDbContext.UserPublicationRoles
+            .Where(urr => urr.UserId == userId)
+            .ToListAsync(cancellationToken);
+
+        await ContentDbContext.RequireTransaction(async () =>
+        {
+            await userPublicationInviteRepository.RemoveByUser(
+                email: userEmail,
+                cancellationToken: cancellationToken);
+
+            await RemoveMany(userPublicationRoles, cancellationToken);
+        });
+    }
+
+    private async Task<string> GetUserEmail(Guid userId, CancellationToken cancellationToken)
+    {
+        return (await userRepository.FindById(userId, cancellationToken))!
```

I don't recommend using a null forgiving operator here, as it opens the door for NREs if no user is found.

---

Ah, I think what I was thinking here is that they should be expected to exist - and therefore it should just blow up if they don't (`500`). i.e. when deleting all resources for a user, the requested user should exist. 

I suppose you might expect a `400`, but our current architecture doesn't lend itself easily to returning this here, since the DELETE methods only return an empty `Task`. Not an `ActionResult` `Either` thingy.

Options are:

- Inject a logger, and log the error to the logger, but still `500`
- Silently return, and do nothing (looks like we deleted stuff, but we actually just silently did nothing)
- Start catching custom domain exceptions in the controller layer, and map this one to a `400` (probably not feasible)

Thoughts?

---

Blowing up seems fine to me. I'd be inclined to do something like this so that it is obvious that blowing up is an intentional choice.
```
var user = await userRepository.FindById(...);
if (user is null) throw new Exception("Uh oh");;
return user.Email;
```

I think the usual pattern is to return null if something isn't found from Get* methods. I might be inclined to move "GetUserEmail" to `userRepository`, in which case it should probably return null if no user is found. Or just inline it GetUserEmail. But I am not fussed so feel free to leave it.

---

Agreed to go with the option of injecting a logger to log the exception, but to let the system blow-up (fail-fast) if the user isn't found, as they should be expected to be found in this context.